prov/verbs/src/verbs_ep.c

```diff
@@ -109,7 +108,45 @@ ssize_t vrb_post_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
 			ep->rq_credits_avail += credits_to_give;
 	}
 
-unlock:
+	return ret;
+}
+
+static int vrb_prepost_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
+{
+	struct vrb_recv_wr *save_wr;
+	size_t i;
+
+	assert(ofi_genlock_held(&vrb_ep2_progress(ep)->ep_lock));
+
+	if (wr->next)
+		return -FI_EINVAL;
+
+	save_wr = vrb_alloc_recv_wr(vrb_ep2_progress(ep));
+	if (!save_wr)
+		return -FI_ENOMEM;
+
+	save_wr->wr.wr_id = wr->wr_id;
+	save_wr->wr.next = NULL;
+	save_wr->wr.num_sge = wr->num_sge;
+	for (i = 0; i < wr->num_sge; i++)
+		save_wr->sge[i] = wr->sg_list[i];
+	save_wr->wr.sg_list = save_wr->sge;
+	slist_insert_tail(&save_wr->entry, &ep->prepost_wr_list);
+	return 0;
+}
+
+ssize_t vrb_post_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
+{
+	int ret;
+
+	if (wr->num_sge > vrb_gl_data.def_rx_iov_limit)
+		return -FI_EINVAL;
```

Comment:
I'd say just assert here, but can we get the limit that was associated with the ep?

Reply:
I would prefer not assert here: `wr->num_sge` is set by the application when the receive buffer is posted: I think it's better to return `-FI_EINVAL` and let the application deal with this error. 

I slightly modified the code based on your suggestion to use the limit that is associated with the ep and I added an assertion failure to ensure that the EP limit is effectively lower than `vrb_gl_data.def_rx_iov_limit`
```
assert(ep->info_attr.rx_iov_limit <= vrb_gl_data.def_rx_iov_limit); 
if (wr->num_sge > ep->info_attr.rx_iov_limit)                       
    return -FI_EINVAL;                                              
```

Reply:
I'm not sure the assert is always true.  The requested limit is subject to the device limits, not necessarily the default limit.

Reply:
The problem is that the pool of WRs belongs to the progress object (shared between EPs of the same domain) and the max number of SGEs is set to `vrb_gl_data.def_rx_iov_limit`. Also, note that a similar assert() was added by the PR at line 68 (same file).
One possible fix would be to detect the max rx_iov_limit set by the HCA and use this value when the pool of WRs is created.

Reply:
Hmm... I believe ibv_post_recv() checks num_sge.  So, I think we're okay moving this check into vrb_prepost_recv(), which is where we really need it.  I agree if we can use the device limit when initializing the pool, that should cover (unlikely) edge cases.  The check in vrb_prepost_recv() just need to verify that we don't overrun the local array.  When the receives are actually posted to the device, vrb_post_recv_internal() -> ibv_post_recv() will check that the QP can handle it.  A failure there is a user error, which at least will be handled (by tearing down the QP).  Does this work for you?